File: src/main/java/com/machfour/macros/storage/SqlUtils.java
package com.machfour.macros.storage;

import com.machfour.macros.core.MacrosPersistable;
import com.machfour.macros.data.*;
import com.sun.istack.internal.Nullable;

import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.*;

class SqlUtils {
    private SqlUtils() {
    }

    static <E> String join(String sep, Iterable<E> toJoin) {
        return join(sep, toJoin.iterator(), null);
    }

    static <E> String join(String sep, Iterator<E> it) {
        return join(sep, it, null);
    }

    static <E> String join(String sep, Iterable<E> toJoin, @Nullable String elementSuffix) {
        return join(sep, toJoin.iterator(), elementSuffix);
    }

    static <E> String join(String sep, Iterator<E> it, @Nullable String elementSuffix) {
        StringBuilder joined = new StringBuilder();
        if (it.hasNext()) {
            joined.append(it.next());
            if (elementSuffix != null) {
                joined.append(elementSuffix);
            }
            while (it.hasNext()) {
                joined.append(sep);
                joined.append(it.next());
                if (elementSuffix != null) {
                    joined.append(elementSuffix);
                }
            }
        }
        return joined.toString();
    }

    private static String makeQuestionMarks(String sep, int howMany) {
        Iterator<Character> questionIterator = new Iterator<Character>() {
            int current = 0;

            @Override
            public boolean hasNext() {
                return current < howMany;
            }

            @Override
            public Character next() {
                Character c = hasNext() ? '?' : null;
                current++;
                return c;
            }
        };
        return join(sep, questionIterator);
    }

    // " WHERE column1 = ?"
    // " WHERE column1 IN (?, ?, ?, ...?)"
    // if nkeys is <= 0, no where string will be formed, so all objects will be returned.
    private static String makeWhereString(Column<?, ?, ?> whereColumn, int nValues) {
        assert nValues >= 0;
        switch (nValues) {
            case 0:
                return "";
            case 1:
                // TODO is this needed?
                //if (whereColumn.type().equals(DATESTAMP)) {
                //    sb.append("DATE(").append(whereColumn.sqlName()).append(")");
                //} else {
                //sb.append(whereColumn.sqlName());
                //}
                return " WHERE " + whereColumn.sqlName() + " = ?";
            default:
                return " WHERE " + whereColumn.sqlName() + " IN (" + makeQuestionMarks(", ", nValues) + ")";
        }
    }

    static <M> String deleteTemplate(Table<M> table, Column<M, ?, ?> whereColumn) {
        return deleteTemplate(table) + makeWhereString(whereColumn, 1);

    }
    // delete all!
    static <M> String deleteTemplate(Table<M> table) {
        return "DELETE FROM " + table.name();
    }

    // " WHERE (likeColumn[0] LIKE likeValue[0]) OR (likeColumn[1] LIKE likeValue[1]) OR ..."
    private static <M> String makeWhereLikeString(List<Column<M, Types.Text, String>> likeColumns) {
        switch (likeColumns.size()) {
            case 0:
                return "";
            case 1:
                return " WHERE " + likeColumns.get(0).sqlName() + " LIKE ?";
            default:
                List<String> bracketedWhereClauses = new ArrayList<>(likeColumns.size());
                for (Column<?, Types.Text, String> c : likeColumns) {
                    bracketedWhereClauses.add("(" + c.sqlName() + " LIKE ? )");
                }
                return join(" OR ", bracketedWhereClauses);
        }
    }

    static <M, T extends MacrosType<J>, J> String selectLikeTemplate(
            Table<M> t, Column<M, ?, ?> selectColumn, List<Column<M, Types.Text, String>> likeColumns) {
        return selectTemplate(t, toList(selectColumn), makeWhereLikeString(likeColumns), false);
    }

    static <M, T extends MacrosType<J>, J> String selectTemplate(
            Table<M> t, Column<M, ?, ?> selectColumn, Column<M, ?, ?> whereColumn, int nValues, boolean distinct) {
        return selectTemplate(t, toList(selectColumn), whereColumn, nValues, distinct);
    }

    private static <M, T extends MacrosType<J>, J> String selectTemplate(
            Table<M> t, List<Column<M, ?, ?>> orderedColumns, String whereString, boolean distinct) {
        List<String> words = new ArrayList<>(6);
        words.add("SELECT");
        if (distinct) {
            words.add("DISTINCT");
        }
        words.add(join(", ", orderedColumns));
        words.add("FROM");
        words.add(t.name());
        words.add(whereString);
        return join(" ", words);
    }

    static <M, T extends MacrosType<J>, J> String selectTemplate(
            Table<M> t, List<Column<M, ?, ?>> orderedColumns, Column<M, T, J> whereColumn, int nValues, boolean distinct) {
        return selectTemplate(t, orderedColumns, makeWhereString(whereColumn, nValues), distinct);
    }

    // columns must be a subset of table.columns()
    static <M> String insertTemplate(Table<M> t, List<Column<M, ?, ?>> orderedColumns) {
        String placeholders = makeQuestionMarks(", ", orderedColumns.size());
        return "INSERT INTO " + t.name() + " (" + join(", ", orderedColumns) + ") VALUES ( " + placeholders + ")";
    }

    static <M, T extends MacrosType<J>, J> String updateTemplate(Table<M> t, List<Column<M, ?, ?>> orderedColumns, Column<M, T, J> keyCol) {
        return "UPDATE " + t.name() + " SET " + join(",", orderedColumns, "= ?") + makeWhereString(keyCol, 1);
    }

    static <M> void bindData(PreparedStatement p, ColumnData<M> values, List<Column<M, ?, ?>> orderedColumns, Object... extras) throws SQLException {
        int colIndex = 1; // parameters are 1 indexed!
        for (Column<M, ?, ?> col : orderedColumns) {
            // Internally, setObject() relies on a ladder of instanceof checks
            p.setObject(colIndex, columnDataToRaw(values, col));
            colIndex++;
        }
        bindObjects(p, colIndex, extras);
    }

    private static void bindObjects(PreparedStatement p, int startIndex, Object... objects) throws SQLException {
        int colIndex = startIndex;
        for (Object o : objects) {
            p.setObject(colIndex, o);
            colIndex++;
        }
    }

    static <E> void bindObjects(PreparedStatement p, List<E> objects) throws SQLException {
        bindObjects(p, 1, objects);
    }

    private static void bindObjects(PreparedStatement p, int startIndex, List<?> objects) throws SQLException {
        int colIndex = startIndex;
        for (Object o : objects) {
            p.setObject(colIndex, o);
            colIndex++;
        }
    }

    static <M, T extends MacrosType<J>, J> Object columnDataToRaw(ColumnData<M> c, Column<M, T, J> col) {
        MacrosType<J> type = col.type();
        return type.toRaw(c.unboxColumn(col));
    }

    static <M, T extends MacrosType<J>, J> void rawToColumnData(ColumnData<M> c, Column<M, T, J> col, Object data) {
        MacrosType<J> type = col.type();
        c.putData(col, type.fromRaw(data));
    }

    static <M extends MacrosPersistable> Map<Long, M> makeIdMap(List<M> objects) {
        Map<Long, M> idMap = new HashMap<>(objects.size(), 1);
        for (M m : objects) {
            idMap.put(m.getId(), m);
        }
        return idMap;
    }

    static <E> List<E> toList(E e) {
        return Collections.singletonList(e);
    }
}

File: src/main/java/com/machfour/macros/sql/SqlUtils.kt
package com.machfour.macros.sql

import com.machfour.macros.core.MacrosEntity
import com.machfour.macros.sql.datatype.MacrosType
import com.machfour.macros.sql.datatype.Types
import com.machfour.macros.sql.generator.ColumnExpr
import com.machfour.macros.sql.generator.Conjuction
import com.machfour.macros.util.stringJoin
import java.io.BufferedReader
import java.io.IOException
import java.io.Reader
import java.sql.SQLException

object SqlUtils {
    internal fun <M> joinColumns(columns: Iterable<Column<M, *>>, suffix: String = ""): String {
        return stringJoin(columns, sep = ", ", itemSuffix = suffix) { it.sqlName }
    }

    // usually, an SQL placeholder is just a question mark. But, for example a DateStamp
    // needs to be entered as DATE('date_string'), so that logic is implemented here
    private fun getSqlPlaceholder(columnType: MacrosType<*>): String {
        // for now, DateStamp is the only special one
        return when (columnType) {
            // need the space here in order for the JDBC binding code to work properly
            Types.DATESTAMP -> "DATE( ? )"
            else -> "?"
        }
    }

    private fun <M> makePlaceholders(columns: List<Column<M, *>>, placeholderFormat: (c: Column<M, *>) -> String): String {
        return stringJoin(columns, sep = ", ") { placeholderFormat(it) }
    }

    // creates SQL placeholders for the given (ordered) list of columns
    private fun <M> makeInsertPlaceholders(columns: List<Column<M, *>>): String {
        return makePlaceholders(columns) { getSqlPlaceholder(it.type) }
    }

    // creates SQL placeholders for the given (ordered) list of columns
    internal fun <M> makeUpdatePlaceholders(columns: List<Column<M, *>>): String {
        return makePlaceholders(columns) { it.sqlName + " = " + getSqlPlaceholder(it.type) }
    }

    // " WHERE column1 = ?"
    // " WHERE column1 IN (?, ?, ?, ...?)"
    // " WHERE column1 IS [NOT] NULL
    // if nkeys is 0, no where string will be formed, so all objects will be returned.
    // exception thrown if nValues < 0
    internal fun makeWhereString(whereColumnExpr: ColumnExpr<*, *>, nValues: Int? = null, isNotNull: Boolean? = null): String {
        val colName = whereColumnExpr.sql
        require((nValues != null) xor (isNotNull != null)) { "Must specify one of nValues or isNull" }

        if (isNotNull != null) {
            val not = if (isNotNull) "NOT" else ""
            return " WHERE $colName IS $not NULL"
        }

        check(nValues != null && nValues >= 0)
        if (nValues == 0) {
            return ""
        }

        val placeholder = getSqlPlaceholder(whereColumnExpr.type)

        //if (whereColumn.type().equals(DATESTAMP)) {
        //    sb.append("DATE(").append(whereColumn.sqlName()).append(")");
        //} else {
        //sb.append(whereColumn.sqlName());
        //}
        return if (nValues == 1) {
            " WHERE $colName = $placeholder"
        } else {
            val placeholders = stringJoin(listOf(placeholder), sep = ", ", copies = nValues)
            " WHERE $colName IN (${placeholders})"
        }
    }

    fun <M> deleteWhereTemplate(table: Table<M>, whereColumn: Column<M, *>, nValues: Int): String {
        return deleteAllTemplate(table) + makeWhereString(whereColumn, nValues)
    }

    fun <M> deleteWhereNullTemplate(table: Table<M>, whereColumn: Column<M, *>, isNotNull: Boolean): String {
        return deleteAllTemplate(table) + makeWhereString(whereColumn, isNotNull = isNotNull)
    }

    // delete all!
    fun <M> deleteAllTemplate(table: Table<M>): String {
        return "DELETE FROM " + table.name
    }

    // " WHERE (likeColumn[0] LIKE likeValue[0]) OR (likeColumn[1] LIKE likeValue[1]) OR ..."
    internal fun <M> makeWhereLikeString(
        likeColumns: Collection<Column<M, String>>,
        conjunction: Conjuction = Conjuction.OR,
    ): String {
        return when (likeColumns.size) {
            0 -> ""
            1 -> " WHERE " + likeColumns.first().sqlName + " LIKE ?"
            else -> {
                val bracketedWhereClauses: MutableList<String> = ArrayList(likeColumns.size)
                for (c in likeColumns) {
                    bracketedWhereClauses.add("(" + c.sqlName + " LIKE ?)")
                }
                " WHERE " + stringJoin(bracketedWhereClauses, sep = " ${conjunction.sql} ")
            }
        }
    }

    // columns must be a subset of table.columns()
    fun <M> insertTemplate(t: Table<M>, orderedColumns: List<Column<M, *>>): String {
        val placeholders = makeInsertPlaceholders(orderedColumns)
        val columnSql = joinColumns(orderedColumns)
        return "INSERT INTO ${t.name} ($columnSql) VALUES ($placeholders)"
    }

    fun <M, J> updateTemplate(t: Table<M>, orderedColumns: List<Column<M, *>>, keyCol: Column<M, J>): String {
        val columnPlaceholders = makeUpdatePlaceholders(orderedColumns)
        val whereString = makeWhereString(keyCol, 1)
        // TODO dynamic placeholders
        return "UPDATE ${t.name} SET $columnPlaceholders $whereString"
    }


    fun <M : MacrosEntity<M>> makeIdMap(objects: Collection<M>): Map<Long, M> {
        val idMap: MutableMap<Long, M> = LinkedHashMap(objects.size, 1.0f)
        objects.forEach {
            val id : Long = it.id
            require(!idMap.containsKey(id)) { "Two objects had the same ID" }
            idMap[id] = it
        }
        return idMap
    }

    // Returns a list of strings such that each one is a complete SQL statement in the SQL
    // file represented by reader r.
    // IMPORTANT DETAIL:
    // It's not enough to do this just by splitting on semicolons, since there are things
    // like nested statements with 'END;' and strings may contain semicolons.
    // So the split token is the semicolon followed by two blank lines, i.e. statements
    // in the SQL file must be terminated by a semicolon immediately followed by \n\n
    // XXX Possible bug if the newline character is different on different platforms.
    @Throws(IOException::class)
    fun createSplitStatements(r: Reader): List<String> {
        return createStatements(r, "\n")
            .split(";\n\n")
            // a double newline after the last statement will cause an empty final element
            // in the split list, which turns into a null statement ';' after the map.
            // The Android SQL library seems to spew at this (throws SQL exception despite the
            // SQL code being 0 (NOT_AN_ERROR).
            // Anyway let's just filter everything to ensure we don't end up with any null statements
            .filter { it.isNotBlank() }
            // replace remaining newline characters by spaces and add the semicolon back
            .map { it.replace("\n".toRegex(), " ") + ";" }
    }

    @Throws(IOException::class)
    fun createStatements(r: Reader, lineSep: String = " "): String {
        val trimmedDecommentedLines = ArrayList<String>()
        BufferedReader(r).use { reader ->
            // steps: remove all comment lines, trim, join, split on semicolon
            while (reader.ready()) {
                var line = reader.readLine()
                val commentIndex = line.indexOf("--")
                if (commentIndex == 0) {
                    continue // skip comment lines completely
                } else if (commentIndex != -1) {
                    line = line.substring(0, commentIndex)
                }
                line = line.replace("\\s+".toRegex(), " ")
                if (line != " ") {
                    trimmedDecommentedLines.add(line)
                }
            }
        }
        return stringJoin(trimmedDecommentedLines, sep = lineSep)
    }

    // Returns triggers that update the create and modify times of the given table.
    // Columns are determined using the Schema object so ensure it is updated.

    fun <M> createInitTimestampTriggers(t: Table<M>): List<String> {
        val table = t.name
        val createTime = t.createTimeColumn.sqlName
        val modifyTime = t.modifyTimeColumn.sqlName
        val id = t.idColumn.sqlName
        val currentUnixTime = "CAST(strftime('%s', 'now') AS INTEGER)"
        val timestampCols = setOf(t.createTimeColumn, t.modifyTimeColumn)
        val columnNames = t.columns.filterNot { it in timestampCols }.map { it.sqlName }

        val initTimeStampTrigger = """
            |CREATE TRIGGER init_${table.lowercase()}_timestamp
            |AFTER INSERT ON $table
            |WHEN (NEW.${createTime} = 0)
            |BEGIN UPDATE $table
            |SET $createTime = $currentUnixTime, $modifyTime = $currentUnixTime 
            |WHERE $id = NEW.${id};
            |END;
        """.trimMargin("|")

        val updateTimestampTrigger = """
            |CREATE TRIGGER update_${table.lowercase()}_timestamp
            |AFTER UPDATE OF ${stringJoin(columnNames, sep = ", ")} ON $table
            |BEGIN UPDATE $table
            |SET $modifyTime = $currentUnixTime
            |WHERE $id = NEW.${id};
            |END;
        """.trimMargin("|")
        return listOf(initTimeStampTrigger, updateTimestampTrigger)
    }

    @Throws(SQLException::class)
    fun <M, J> rethrowAsSqlException(rawValue: Any?, c: Column<M, J>) {
        throw SQLException("Could not convert value '$rawValue' for column ${c.table}.${c.sqlName} (type ${c.type})")
    }
}